Add the option to use the helix slope to decide the LoopHelixFit direction hypothesis

michaelmackenzie opened the pull request:
This adds the option to use the helix seed slope's significance to decide the fit direction (downstream vs. upstream) to assume in the LoopHelixFit. In the case the slope isn't very significant, both fit directions are fit and the p(chi^2) results are compared. The goal of this option is to fit downstream and upstream tracks in a single LoopHelixFit instance, producing a single output track collection where each helix seed corresponds to at most one track direction hypothesis (there can still be different track mass hypothesis collections).

In the ambiguous fit direction case where a calo-hit is included in one fit result but not the other hypothesis fit result, the fit that included the calo-hit is taken instead. This recovers a significant fraction of incorrectly selected CE- and CE+ direction tracks without additional loss.

I additionally added the option to cut on the helix slope significance when fitting a fixed direction hypothesis, skipping helix seeds with a helix slope that is unlikely to correspond to the fit direction hypothesis. This configuration can then be used in the trigger, to not waste time fitting tracks that are unlikely to match the fit hypothesis.

The current update does not change and default behavior. The next step would be investigating making CPR more fit hypothesis agnostic, which a Yale undergrad will look into, at which point we would be able to use this update to merge the De (Dmu) and Ue (Umu) reconstruction paths into a single Electron (Muon) reconstruction path.

An example fcl file to process events with this configuration is here:
`/exp/mu2e/app/users/mmackenz/Yale/trigger/helix_slope/test.fcl`

FNALbuild commented:
Hi @michaelmackenzie,
You have proposed changes to files in these packages:
- RecoDataProducts
- Mu2eKinKal

which require these tests: build.

@Mu2e/write, @Mu2e/fnalbuild-users have access to CI actions on [main](/Mu2e/Offline/tree/main).





:hourglass: The following tests have been triggered for db6b6b9a2e7dbc4b3848a4475a1cb053ee6e6462: build  (Build queue has 1 jobs)


[About FNALbuild.](https://mu2ewiki.fnal.gov/wiki/GitHubWorkflow#GitHub_Pull_Request_Procedures_and_FNALbuild) [Code review on Mu2e/Offline.](https://mu2ewiki.fnal.gov/wiki/GitHubWorkflow#Code_Review)

FNALbuild commented:
:sunny: The build tests passed at db6b6b9a2e7dbc4b3848a4475a1cb053ee6e6462.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| test with | :white_check_mark: | Command did not list any other PRs to include |
| merge | :white_check_mark: | Merged db6b6b9a2e7dbc4b3848a4475a1cb053ee6e6462 at df7782c2fb9cace1b15a10c8574e3317d226fc80 |
| build (prof) | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/scons.log). Build time: 09 min 40 sec |
| ceSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/ceSimReco.log)  |
| g4test_03MT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/g4test_03MT.log)  |
| transportOnly | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/transportOnly.log)  |
| POT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/POT.log)  |
| g4study | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/g4study.log)  |
| cosmicSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/cosmicSimReco.log)  |
| cosmicOffSpill | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/cosmicOffSpill.log)  |
| ceSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/ceSteps.log)  |
| ceDigi | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/ceDigi.log)  |
| muDauSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/muDauSteps.log)  |
| ceMix | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/ceMix.log)  |
| rootOverlaps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/rootOverlaps.log)  |
| g4surfaceCheck | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/g4surfaceCheck.log)  |
| trigger | :x: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/trigger.log) Return Code 2. |
| FIXME, TODO | :large_orange_diamond: | [TODO (6) FIXME (0) in 3 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/fixme_todo.log) |
| clang-tidy | :large_orange_diamond: | [0 errors 549 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/clang-tidy.log) |
| whitespace check | :white_check_mark: | [no whitespace errors found](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/artifact/whitespace_errs.log) |


N.B. These results were obtained from a build of this Pull Request at db6b6b9a2e7dbc4b3848a4475a1cb053ee6e6462 after being merged into the base branch at df7782c2fb9cace1b15a10c8574e3317d226fc80.


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2377/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

brownd1978 reviewed (changes requested):
Hi Michael this is good and important work.
As I see it there are 3 separate functionalities introduced in this PR. I would like to see them separated more in code. The separable functions I see are:
- A function to decide if a helix fit direction measurement is significant. This has nothing directly to do with LoopHelixFit and should be implemented external to the module.
-  compare 2 LoopHelix fits of the same hits with opposite physical direction and classify them as clearly upstream, clearly downstream, or ambiguous. This could be implemented as an external helper function, making it easy to try different algorithms.
- a logical function to decide which fits to create and which to keep the using the above functions, plus the option of forcing the direction by hand. Probably this works best as a LoopHelixFit member function.
I think splitting these functions out will make the code clearer, more flexible and more maintainable.

michaelmackenzie commented:
Hi Dave,

Thank you for the feedback! I agree we should aim for cleanliness and modularity here.

> * A function to decide if a helix fit direction measurement is significant. This has nothing directly to do with LoopHelixFit and should be implemented external to the module.

This is definition of fit direction ambiguity is defined by the `HelixRecoDir` helix seed field, but I was worried different helix collections may be using different values and that one may want to chose a different significance value without needing to recreate the helix seed collection. The helix slope significance isn't calculated here, it's stored in the helix seed, so the only evaluation being done here is comparing to the configured decision threshold. Is the preferred mode inheriting the sigma threshold from the helix seed collections, accepting that this could differ between collections and that using different thresholds will require unique helix seed collections in addition to unique LoopHelixFit instances?

> * compare 2 LoopHelix fits of the same hits with opposite physical direction and classify them as clearly upstream, clearly downstream, or ambiguous. This could be implemented as an external helper function, making it easy to try different algorithms.

By this do you mean add a helper function outside of LoopHelixFit that compares two KKTrack<KinKal::LoopHelix> objects and decides which is the better fit? I think this makes sense and I'll get started, though I would still advocate for the preference of tracks with calo-hits as it recovers a large fraction of incorrectly fit conversion track directions, but this may not be as general if comparing two downstream or two upstream tracks in a different use case. Where is the best place for this helper function to live?

> * a logical function to decide which fits to create and which to keep the using the above functions, plus the option of forcing the direction by hand. Probably this works best as a LoopHelixFit member function.

By this do you mean essentially moving the code within the for-loop to a member function that is called to make a specific track fit hypothesis (upstream vs. downstream in this case, but this could be extended to mass hypothesis etc.)? I can do this, and I agree separating out this logic should be cleaner.

>   I think splitting these functions out will make the code clearer, more flexible and more maintainable.

I agree, I'll work on splitting and cleaning the logic I've added here to be clearer and more maintainable (and more flexible if I can!).

FNALbuild commented:
:memo: The HEAD of `main` has changed to 5e277911e8733d0c542c8ade6afdddfb5c7c879d. Tests are now out of date.

brownd1978 commented:
> By this do you mean add a helper function outside of LoopHelixFit that compares two KKTrackKinKal::LoopHelix objects and decides which is the better fit?
 Yes. If you first define a virtual function it could have several implementaitons. Parameters should be passed via fcl. There are several examples how to do this in LoopHelixModule/KKFit.

brownd1978 commented:
> By this do you mean essentially moving the code within the for-loop to a member function
Yes.

brownd1978 commented:
> The helix slope significance isn't calculated here, it's stored in the helix seed
Even if currently this selection is just a comparison with a data member, it's good to buffer the selection through a function, to allow you to possibly extend the test in future without changing the code structure.

brownd1978 commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -318,12 +339,43 @@ namespace mu2e {
+    const bool goodfit_1(goodFit(ktrk_1)), goodfit_2(goodFit(ktrk_2));
+    if(print_>1) printf(" [LoopHelixFit::%s]: status 1 = %o, status 2 = %o\n", __func__, goodfit_1, goodfit_2);
+
+    if(!goodfit_2) return true; //if the second is bad or if both are bad, default to the first
```
If the function returns on this statement the first fit could still be bad.
Returning a more complicated return value than 'bool' would address this, by having 3 states (1 is better, 2 is better, neither are good)

michaelmackenzie replied:
If both fits are bad, this is caught outside this comparison in the requirement that the accepted track is a good fit. In the case all fits are meant to be saved, this defaults to saving the first track given to the comparison. I updated to return 1 if the first is better, 0 if the second is better, and -1 if both are bad fits, still defaulting to the first track if both are bad.

brownd1978 commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -318,12 +339,43 @@ namespace mu2e {
+
+    const auto p_1(ktrk_1.fitStatus().chisq_.probability());
+    const auto p_2(ktrk_2.fitStatus().chisq_.probability());
+    const auto n_calo_1(ktrk_1.caloHits().size());
```
The indenting looks odd here, please check

michaelmackenzie replied:
I'm not sure if I can see the issue, it looks like it matches the other functions' indenting.

brownd1978 commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -318,12 +339,43 @@ namespace mu2e {
+    if(print_>1) printf(" [LoopHelixFit::%s]: p 1 = %.4f, n(calo-hits) 1 = %lu, p 2 = %.4f, n(calo-hits) 2 = %lu\n",
+                        __func__, p_1, n_calo_1, p_2, n_calo_2);
+
+    if(n_calo_1 == n_calo_2) return p_1 >= p_2; //if they both have the same number of calo-hits, use p(chi^2)
```
I suggest making the calo hit comparison optional and configurable. It's not clear to me how this responds to true upstream tracks for instance.

michaelmackenzie replied:
Processing cosmic events with upstream and downstream tracks, including this prioritization of tracks with calo-hits also improved the true selection efficiency for both true propagation directions (but not by as much as for CEs). I've made this a configurable flag, defaulting to true since it improves the selection efficiency for all cases I've checked so far (CE, downstream cosmics, and upstream cosmics).

brownd1978 commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -490,6 +616,26 @@ namespace mu2e {
+      const float slopeSig = (slopeErr > 0.f) ? slope/slopeErr : 0.f;
+      if(slopeSig*fdir_.dzdt() < slopeSigCut_) { //slope is below the given threshold relative to the given direction
+        if(print_ > 0) printf("[LoopHelixFit::%s] Skipping helix seed with slope significance %.1f\n", __func__, slopeSig);
+        return false;
```
I am confused by the logic: if the slope isn't significant, don't we still want to fit the track? Especially if the direction is hard-coded.

michaelmackenzie replied:
This was a request from Giani, that for the trigger we may want to skip fitting helix seeds that are very likely upstream tracks when only looking for downstream tracks. In the Offline processing, I imagine there's no harm in fitting these downstream as well since we're forcing a downstream fit for all tracks.

brownd1978 replied:
I think the intent is:
- if the slope is significant, only fit that one (significant) direction
- if the slope isn't significant fit both directions
Maybe I've misunderstood the logic, but it seems the fit is skipped entirely if the slope isn't significant.

michaelmackenzie replied:
This would only be used if using a forced fit direction, and throws an error if used in the helix slope-based fit direction mode. The intent with this is in the Online processing, only fit with the forced downstream hypothesis and skip events that have a significant upstream slope to save processing time with minimal efficiency loss for downstream tracks. This update is essentially separate from the rest of the updates as it's only relevant to the forced fit direction processing

brownd1978 replied:
I'm starting to think this test belongs outside loopHelixFit, either in the helix fitter itself, or (after merging) in a dedicated module that would only be run in the trigger reco sequence.

michaelmackenzie replied:
That's a good point. Perhaps this should be a new producer module that reads in a helix seed collection and produces a filtered helix collection? This way you can tailor the collection for the LoopHelixFit input, but not lose these helices if another path may want to use them in a separate LoopHelixFit module instance (e.g. an upstream trigger that only looks at significantly upstream-oriented helices using the same initial helix collection).

michaelmackenzie replied:
I moved this filtering to the HelixFilter module, which seemed a more appropriate place for this type of selection.

brownd1978 commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -347,88 +475,86 @@ namespace mu2e {
-            fitpart = static_cast<PDGCode::type>(-1*fitpart); // reverse sign
+
+        // determine the fit direction hypothesis
+        auto helix_dir = chooseHelixDir(hseed);
```
I recommend combining the 2 hseed functions into 1, returning a vector of fit directions. That could be empty (bad helix), have 1 entry, or have 2 entries. Then you don't have to worry about interactions between these functions.

michaelmackenzie replied:
I'm not positive I follow this suggestion exactly. Is it that I return a vector of directions, loop over this filling a vector of track hypotheses, and then choose the best hypothesis from this list (which ideally should be length 1 or 2, where 0 means the helix seed was a bad seed)? If my understanding is right I can start making this change.

brownd1978 replied:
>  If my understanding is right I can start making this change.
That's what I meant.
The goal is to make the code executed in produce as simple as possible

michaelmackenzie replied:
I made this update, where I return a vector of length 0, 1, or 2 and loop over this vector and perform a track fit for each hypothesis. If ever extended, it will keep running `compare_tracks` taking the best result from the loop for any number of hypotheses (but of course there are only two at maximum as of now).

brownd1978 commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -347,88 +475,86 @@ namespace mu2e {
+          const bool use_downstream = !ktrk_up || (ktrk_down && compare_tracks(*ktrk_down, *ktrk_up)); //if one is null, default to the other (use down if both are null, doesn't matter)
+          if(print_ > 0)
+            printf("[LoopHelixFit::%s]: Fit both direction options, slope = %9.2e, sig = %.2f, status down = %i, status up = %i, p(chi^2) down = %.4f, p(chi^2) up = %.4f: chose %s\n",
+                   __func__, hseed.recoDir().slope(), hseed.recoDir().slopeSig(),
```
This statement is too complex. Please recode it as a function.

>

michaelmackenzie replied:
I've updated to pass the pointers rather than the references, so the `compare_tracks` function can also check if the pointer is null (though I'm not as used to working with `std::unique_ptr` so let me know if I haven't used it correctly)

brownd1978 commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -347,88 +475,86 @@ namespace mu2e {
+        if(print_>1) ktrk->printFit(std::cout,print_-1);
+
+        // save the fit result
+        if(goodfit || saveall_){
```
since goodfit has been tested several times already before getting to this statement I think the saveall_ flag no longer does anything. We should discuss whether to remove it or redefine it.

michaelmackenzie replied:
I think with `saveall_` set it will still save the track bad fit track (possibly without the drift fit or extrapolation performed) as was true before. In the case both fits are performed and both are bad, the downstream is saved with this flag. I don't know what this feature is used for, so I'm open to removing or redefining it in anyway that is suggested!

brownd1978 replied:
saveall_ is for debugging failed fits. It's never used in production.

michaelmackenzie commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -155,7 +156,9 @@ namespace mu2e {
-    fhicl::Atom<int> fitDirection { Name("FitDirection"), Comment("Particle direction to fit, either upstream or downstream") };
+    fhicl::OptionalAtom<double> slopeSigThreshold{ Name("SlopeSigThreshold"), Comment("Helix slope significance threshold to assume the direction")};
+    fhicl::OptionalAtom<double> slopeSigCut{ Name("SlopeSigCut"), Comment("Helix slope significance cut when assuming a fit direction")};
+    fhicl::Atom<int> fitDirection { Name("FitDirection"), Comment("Particle direction to fit, either upstream or downstream"), [this](){ return !slopeSigThreshold(); }};
```
I think this is similar to the current setup, so I don't think the downstream logic changes. In this setup, it throws an error if one tries to set `SlopeSigThreshold` and `FitDirection` saying `FitDirection` isn't a supported parameter. I updated to an OptionalAtom so now we can better control the error message.

michaelmackenzie commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -155,7 +156,9 @@ namespace mu2e {
     fhicl::OptionalTable<KKExtrapConfig> Extrapolation { Name("Extrapolation") };
     // LoopHelix module specific config
-    fhicl::Atom<int> fitDirection { Name("FitDirection"), Comment("Particle direction to fit, either upstream or downstream") };
+    fhicl::OptionalAtom<double> slopeSigThreshold{ Name("SlopeSigThreshold"), Comment("Helix slope significance threshold to assume the direction")};
```
I've updated the SlopeSigCut name to SlopeSigVetoCut to make it clearer that this parameter is to cut out helix seeds. I've also updated the comments to say "Input helix seed..." to be clear that these values are applied to the inputs.

michaelmackenzie commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -318,12 +339,43 @@ namespace mu2e {
     if(print_ > 0) kkbf_->print(std::cout);
+
+    // Print the fit direction configuration
+    if(print_ > -1) printf("[LoopHelixFit::%s::%s] Fit dz/dt direction = %.0f, PDG = %i, use helix slope = %o with threshold %.1f, use helix slope cut = %o with theshold %.1f\n",
```
I've changed this to `print_ > 0`, as I only set this to `-1` to get this summary info for my larger tests. In general I use `print_ > -1` checks for printout that should be there in normal running, but can be silenced if needed for some reason (but I don't think this is the case here).

michaelmackenzie commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -318,12 +339,43 @@ namespace mu2e {
+    if(useHelixSlope_) return hseed.recoDir().predictDirection(slopeSigThreshold_);
+    if(fdir_.fitDirection() == TrkFitDirection::FitDirection::upstream  ) return HelixRecoDir::PropDir::upstream;
+    if(fdir_.fitDirection() == TrkFitDirection::FitDirection::downstream) return HelixRecoDir::PropDir::downstream;
+    return HelixRecoDir::PropDir::ambiguous;
```
I don't think it can ever reach the final return statement, I just wanted to protect against any unexpected cases (perhaps a forced TrkFitDirection that is ambiguous and forces the track fit to do both fits each time?). I can remove this case or throw an error here instead if this is better.

michaelmackenzie commented:
I rebased and force-pushed the updates to not conflict with the HelixFilter updates at the HEAD of Offline. I apologize if this causes any issues with the review comments!

brownd1978 commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -318,12 +342,53 @@ namespace mu2e {
+  }
+
+  int LoopHelixFit::compare_tracks(const std::unique_ptr<KKTRK>& ktrk_1, const std::unique_ptr<KKTRK>& ktrk_2) const {
+    // return 1 to select the first track, 0 to select the second, -1 to indicate both are bad tracks
```
Please code these return values as a (private) enum, so that downstream interpretation will be self-evident;

michaelmackenzie replied:
I updated this comparison to use the private enums `kFirstPass`, `kSecondPass`, and `kBothFail` to be more clear.

brownd1978 commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -347,88 +488,103 @@ namespace mu2e {
+              ktrk = std::move(ktrk_i);
+              helix_dir = dir;
+            } else {
+              const bool accept = compare_tracks(ktrk, ktrk_i) == 0;
```
This logic would be clearer if you started with an array<std::unique_ptr<KKTRK>,2>, filled that in the loop over helix_dirs, and then called compare_tracks on the array.

michaelmackenzie replied:
I updated to this style, passing the array to the compare tracks call.

brownd1978 commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -165,11 +168,16 @@ namespace mu2e {
+      bool goodHelix(HelixSeed const& hseed) const;
+      std::vector<HelixRecoDir::PropDir> chooseHelixDir(HelixSeed const& hseed) const;
+      int compare_tracks(const std::unique_ptr<KKTRK>& ktrk_1, const std::unique_ptr<KKTRK>& ktrk_2) const;
+      std::unique_ptr<KKTRK> fitTrack(art::Event& event, HelixSeed const& hseed, const float dir, const PDGCode::type fitpart);
```
I'm unclear why you pass dir as a float. Using TrkFitDirection would make the interface unambiguous. If you need the FP direction (line 413) you can get that from TrkFitDirection::dzdt()

michaelmackenzie replied:
I updated to passing the HelixRecoDir::PropDir enum directly, and then checking this enum to decide on the slope direction.

brownd1978 commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -347,88 +488,103 @@ namespace mu2e {
+            fitflag.clear(TrkFitFlag::FitOK);
+          if(undefined_dir) fitflag.merge(TrkFitFlag::AmbFitDir);
+          auto kkseed = kkfit_.createSeed(*ktrk,fitflag,*calo_h);
+          if(print_>0) printf("[LoopHelixFit::%s::%s] Create seed             : fitcon = %.4f, nHits = %2lu, seedActiveHits = %2u, %lu calo-hits\n",
```
these additional print lines make it hard to read the code, for instance finding the sampleFit() call. Please move the prints into a single function.

michaelmackenzie replied:
I agree, I was adding these mostly for the debugging process to understand if I was change any track information by fitting once vs. twice. I've moved these to a `print_track_info` function, which I can remove or cleanup more now that the debugging has settled.

brownd1978 reviewed (changes requested):
Thanks for moving the helix direction FP testing outside this class.
This is getting closer, but I still think there are places where the code needs to be simplified and clarified, see the comments below.

michaelmackenzie commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -253,6 +273,10 @@ namespace mu2e {
         kkbf_ = std::move(std::make_unique<KKConstantBField>(VEC3(0.0,0.0,bz)));
       }
+
+      if(useFitDir_ && useHelixSlope_) //can either force the direction or determine it from the helix slope
```
These are two different configuration options: force the fit direction to be a given direction for all helices or decide the fit direction on a helix-by-helix basis. In the latter option, it looks at the helix slope to decide if the direction is clear or if it's ambiguous and needs to be fit twice to decide. One could adjust the helix slope threshold to either always guess a direction (0 threshold) or never guess and always fit twice (very large sigma threshold), but these are both in the latter fit mode option.

brownd1978 commented:
It will be faster to talk on zoom to resolve this.

On Fri, Feb 7, 2025 at 10:09 AM michaelmackenzie ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In Mu2eKinKal/src/LoopHelixFit_module.cc
> <https://github.com/Mu2e/Offline/pull/1401#discussion_r1946965559>:
>
> > @@ -253,6 +273,10 @@ namespace mu2e {
>          fixedfield_ = true;
>          kkbf_ = std::move(std::make_unique<KKConstantBField>(VEC3(0.0,0.0,bz)));
>        }
> +
> +      if(useFitDir_ && useHelixSlope_) //can either force the direction or determine it from the helix slope
>
> These are two different configuration options: force the fit direction to
> be a given direction for all helices or decide the fit direction on a
> helix-by-helix basis. In the latter option, it looks at the helix slope to
> decide if the direction is clear or if it's ambiguous and needs to be fit
> twice to decide. One could adjust the helix slope threshold to either
> always guess a direction (0 threshold) or never guess and always fit twice
> (very large sigma threshold), but these are both in the latter fit mode
> option.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/1401#discussion_r1946965559>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH575Y2FRCRNVATJYXSPL2OTZHPAVCNFSM6AAAAABVYNBKMKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMBSGQ3TENJZG4>
> .
> You are receiving this because you are on a team that was mentioned.Message
> ID: ***@***.***>
>


-- 
David Brown            ***@***.***
Office Phone (510) 486-7261
Lawrence Berkeley National Lab
M/S 50R5008 (50-6026C)  Berkeley, CA 94720

brownd1978 reviewed (changes requested):
Thanks for the updates, the code is considerably simpler and clearer. The only comment this update doesn't address is consolidation of HelixRecoDir and TriFitDirection. I have a few follow-up issues and questions as well.

rlcee commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -490,6 +618,16 @@ namespace mu2e {
+    if(!hseed.status().hasAllProperties(goodseed_)) return false;
+    // test helix
+    auto const& helix = hseed.helix();
+    if(helix.radius() == 0.0 || helix.lambda() == 0.0)
```
I think we need to move to a model where the only reason we throw is is the configuration is missing, read or write error, or memory corruption is detected.  We should get used to skipping things like bad fits with appropriate message facility calls.  Data will be messier than sim and we don't want to just quit during pass1 processing if we can still process the next event correctly.  Of course this is ultimately a collaboration decision/project, so I'm just agitating.

michaelmackenzie replied:
I updated this check to simply skip this helix rather than throw an error.

michaelmackenzie commented on TrkFilters/src/HelixFilter_module.cc:
```diff
@@ -122,6 +126,9 @@ namespace mu2e
         float hRatio     = helTool.hitRatio();
+        const float slope    = Helix.recoDir().slope();
+        const float slopeErr = std::fabs(Helix.recoDir().slopeErr());
+        const float slopeSig = (slopeErr > 0.f) ? slope/slopeErr : 0.f;
```
It's intentionally a signed significance, as if we want, for example, potentially downstream tracks we can cut at a significance of -2 sigma (so keeping all downstream-moving helices as well as ones that are not significantly marked upstream), which would only veto helices that are very likely upstream-moving. If we use an unsigned significance, we can't tell if it's upstream- or downstream-moving, only whether or not the slope is well known.

michaelmackenzie replied:
I'll rename the variable to be clearer that it's a two-sided test. For fitting downstream-only tracks, we don't want to cut out helices between -2-2 sigma as these are reasonable candidates for a downstream track. We would only cut out sig. < -2 sigma as these are very likely upstream-moving tracks. Using the unsigned significance cut at 2 sigma we would cut out -2-2 sigma while leaving in > 2 sigma likely upstream-moving tracks with this cut. If we added a field to also select the slope sign, we couldn't keep the -2-0 sigma helices which are reasonable downstream-moving candidates.

michaelmackenzie commented on RecoDataProducts/inc/TrkFitFlag.hh:
```diff
@@ -15,10 +15,12 @@ namespace mu2e {
+    enum bit_type {hitsOK=0,circleOK,phizOK,helixOK,seedOK,kalmanOK,circleInit,phizInit, //7
+    circleConverged,phizConverged,helixConverged,seedConverged,kalmanConverged, //12
+    MatCorr, BFCorr, FitOK, //15
+    KSF=16, KFF, TPRHelix, CPRHelix, Straight, KKLoopHelix,KKCentralHelix,KKLine, APRHelix, MPRHelix, AmbFitDir, //26
```
This flag was requested to mark tracks that had unknown fit directions at the LoopHelixFit stage. The helix fit direction only stores the fit direction estimate at the helix producer stage, whereas the track fitter can use a different significance threshold or use a forced fit direction and ignore the helix slope entirely (e.g. one fit module instance could only fit downstream, and another could use the helix slope direction to determine the fit direction). I have no strong feeling on keeping or dropping this flag as I don't know how useful it is in the end.

michaelmackenzie commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -347,88 +499,64 @@ namespace mu2e {
-      }
-    }
+        // determine the fit to use
+        const int index = (compare_tracks(ktrks) == kSecondPass) ? 1 : 0;
```
I've updated the logic to simply store each "good" track fit (or all if save all is chosen), including both fits if both are reasonable. I've also removed the track comparison function as it's no longer needed as downstream users can decide how to choose.

michaelmackenzie commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -347,88 +499,64 @@ namespace mu2e {
+        for(unsigned index = 0; index < dirs_size; ++index) {
+          const auto helix_dir_i = helix_dirs.at(index);
+          // if using a fixed direction, only fit helices in this direction
+          if(useFitDir_ && !((fdir_ == TrkFitDirection::downstream && helix_dir_i == HelixRecoDir::PropDir::downstream) ||
```
I moved this masking to the `chooseHelixDir` function, where I filter down the list if the fit direction mask is being applied.

michaelmackenzie commented:
I rebased to resolve conflicts with the main branch (b-field protection update). I believe the current version addresses the issues brought up, but please let me know if there are more updates requested! I believe each iteration has led to a cleaner version of the logic, so this review has been very fruitful.

brownd1978 commented on Mu2eKinKal/src/LoopHelixFit_module.cc:
```diff
@@ -333,6 +382,84 @@ namespace mu2e {
+    // compute the charge from the helicity, fit direction, and BField direction
+    double bz = bnom.Z();
+    const int charge = static_cast<int>(copysign(PDGcharge_,(-1)*helix.helicity().value()*bz*fdir.dzdt()));
+    // test consistency.  Modify this later when the HelixSeed knows which direction it's going TODO
```
This comment looks obsolete

brownd1978 reviewed (approved):
This looks good Michael, thanks for all the updates.
I think it's clearer what the module is doing without compareTracks. That functionality can be re-introduced in a [KalSeedSelector](https://github.com/Mu2e/Offline/blob/main/TrkReco/inc/KalSeedSelector.hh
), using [MergeKalSeeds](https://github.com/Mu2e/Offline/blob/main/TrkReco/src/MergeKalSeeds_module.cc) to produce a pointer list reduced to fits that are clearly upstream or downstream. I imagine PID can be handled the same way. That avoids having reconstruction make decisions that might affect or bias a downstream analyzer.

brownd1978 commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for 5d5d63625d11f88906317107d3c487a00256e17b: build  (Build queue is empty)

FNALbuild commented:
:sunny: The build tests passed at 5d5d63625d11f88906317107d3c487a00256e17b.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| test with | :white_check_mark: | Command did not list any other PRs to include |
| merge | :white_check_mark: | Merged 5d5d63625d11f88906317107d3c487a00256e17b at 415d472c53b4ed9542cd6a7a63a2347ca7567c3e |
| build (prof) | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/scons.log). Build time: 04 min 23 sec |
| ceSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/ceSimReco.log)  |
| g4test_03MT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/g4test_03MT.log)  |
| transportOnly | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/transportOnly.log)  |
| POT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/POT.log)  |
| g4study | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/g4study.log)  |
| cosmicSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/cosmicSimReco.log)  |
| cosmicOffSpill | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/cosmicOffSpill.log)  |
| ceSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/ceSteps.log)  |
| ceDigi | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/ceDigi.log)  |
| muDauSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/muDauSteps.log)  |
| ceMix | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/ceMix.log)  |
| rootOverlaps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/rootOverlaps.log)  |
| g4surfaceCheck | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/g4surfaceCheck.log)  |
| trigger | :x: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/trigger.log) Return Code 2. |
| FIXME, TODO | :large_orange_diamond: | [TODO (6) FIXME (1) in 8 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/fixme_todo.log) |
| clang-tidy | :large_orange_diamond: | [0 errors 798 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/clang-tidy.log) |
| whitespace check | :white_check_mark: | [no whitespace errors found](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/artifact/whitespace_errs.log) |


N.B. These results were obtained from a build of this Pull Request at 5d5d63625d11f88906317107d3c487a00256e17b after being merged into the base branch at 415d472c53b4ed9542cd6a7a63a2347ca7567c3e.


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2393/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

michaelmackenzie commented:
@FNALbuild run build test with https://github.com/Mu2e/mu2e_trig_config/pull/60

FNALbuild commented:
:hourglass: The following tests have been triggered for 5a1efa21f3d69e9ddc7c2c7c3935d5d6bd8773e6: build  (Build queue has 1 jobs)

FNALbuild commented:
:umbrella: The build is failing at 5a1efa21f3d69e9ddc7c2c7c3935d5d6bd8773e6.

```
scons: *** [build/al9-prof-e28-p066/Offline/tmp/CalPatRec/src/AgnosticHelixFinder_module.os] Error 1
```


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| test with | :x: | Failed to retrieve request base branch commit sha for Mu2e/mu2e_trig_config#60, default is to merge into main |
| test with | :white_check_mark: | Included Mu2e/mu2e_trig_config#60 @ d04b96f32a4ff66724dd2e447e8097df6337f8ac by merge |
| merge | :white_check_mark: | Merged 5a1efa21f3d69e9ddc7c2c7c3935d5d6bd8773e6 at 415d472c53b4ed9542cd6a7a63a2347ca7567c3e |
| build (prof) | :x: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/scons.log).  |
| ceSimReco | :wavy_dash: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/ceSimReco.log)  |
| g4test_03MT | :wavy_dash: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/g4test_03MT.log)  |
| transportOnly | :wavy_dash: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/transportOnly.log)  |
| POT | :wavy_dash: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/POT.log)  |
| g4study | :wavy_dash: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/g4study.log)  |
| cosmicSimReco | :wavy_dash: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/cosmicSimReco.log)  |
| cosmicOffSpill | :wavy_dash: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/cosmicOffSpill.log)  |
| ceSteps | :wavy_dash: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/ceSteps.log)  |
| ceDigi | :wavy_dash: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/ceDigi.log)  |
| muDauSteps | :wavy_dash: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/muDauSteps.log)  |
| ceMix | :wavy_dash: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/ceMix.log)  |
| rootOverlaps | :wavy_dash: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/rootOverlaps.log)  |
| g4surfaceCheck | :wavy_dash: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/g4surfaceCheck.log)  |
| trigger | :heavy_exclamation_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/trigger.log)  |
| FIXME, TODO | :large_orange_diamond: | [TODO (6) FIXME (5) in 13 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/fixme_todo.log) |
| clang-tidy | :large_orange_diamond: | [20 errors 1446 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/clang-tidy.log) |
| whitespace check | :white_check_mark: | [no whitespace errors found](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/artifact/whitespace_errs.log) |


N.B. These results were obtained from a build of this Pull Request at 5a1efa21f3d69e9ddc7c2c7c3935d5d6bd8773e6 after being merged into the base branch at 415d472c53b4ed9542cd6a7a63a2347ca7567c3e.


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2395/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

michaelmackenzie commented:
@FNALbuild run build test with https://github.com/Mu2e/mu2e_trig_config/pull/60

FNALbuild commented:
:hourglass: The following tests have been triggered for 73a6af17ce2c9228e9fe1871a01665f671c76d7e: build  (Build queue is empty)

FNALbuild commented:
:sunny: The build tests passed at 73a6af17ce2c9228e9fe1871a01665f671c76d7e.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| test with | :x: | Failed to retrieve request base branch commit sha for Mu2e/mu2e_trig_config#60, default is to merge into main |
| test with | :white_check_mark: | Included Mu2e/mu2e_trig_config#60 @ d04b96f32a4ff66724dd2e447e8097df6337f8ac by merge |
| merge | :white_check_mark: | Merged 73a6af17ce2c9228e9fe1871a01665f671c76d7e at 415d472c53b4ed9542cd6a7a63a2347ca7567c3e |
| build (prof) | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/scons.log). Build time: 04 min 23 sec |
| ceSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/ceSimReco.log)  |
| g4test_03MT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/g4test_03MT.log)  |
| transportOnly | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/transportOnly.log)  |
| POT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/POT.log)  |
| g4study | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/g4study.log)  |
| cosmicSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/cosmicSimReco.log)  |
| cosmicOffSpill | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/cosmicOffSpill.log)  |
| ceSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/ceSteps.log)  |
| ceDigi | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/ceDigi.log)  |
| muDauSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/muDauSteps.log)  |
| ceMix | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/ceMix.log)  |
| rootOverlaps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/rootOverlaps.log)  |
| g4surfaceCheck | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/g4surfaceCheck.log)  |
| trigger | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/trigger.log)  |
| FIXME, TODO | :large_orange_diamond: | [TODO (6) FIXME (5) in 13 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/fixme_todo.log) |
| clang-tidy | :large_orange_diamond: | [0 errors 1448 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/clang-tidy.log) |
| whitespace check | :white_check_mark: | [no whitespace errors found](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/artifact/whitespace_errs.log) |


N.B. These results were obtained from a build of this Pull Request at 73a6af17ce2c9228e9fe1871a01665f671c76d7e after being merged into the base branch at 415d472c53b4ed9542cd6a7a63a2347ca7567c3e.


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2397/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

FNALbuild commented:
:memo: The HEAD of `main` has changed to 0e62372664f83586a6e773dba177415a75afc1b9. Tests are now out of date.

rlcee reviewed (approved):
Overall, looks good to me.  I didn't dig into the actual tracking code or algorithm.  Here are a few things to think about for the future.
- can FitDirection use EnumToStringSparse?
- can  print_track_info use routines in Print or migrate to Print?
- can printf be cout?
